src/libsyntax/parse/lexer/mod.rs

```diff
@@ -27,26 +27,47 @@ pub use ext::tt::transcribe::{TtReader, new_tt_reader, new_tt_reader_with_doc_fl
 pub mod comments;
 mod unicode_chars;
 
+pub type TokenAndSpanResult = Result<TokenAndSpan, ()>;
+
 pub trait Reader {
     fn is_eof(&self) -> bool;
-    fn next_token(&mut self) -> TokenAndSpan;
+    fn try_next_token(&mut self) -> TokenAndSpanResult;
+    fn next_token(&mut self) -> TokenAndSpan {
+        match self.try_next_token() {
+            Ok(tok) => tok,
+            Err(_) => {
+                self.emit_fatal_errors();
+                panic!(FatalError);
+            }
+        }
+    }
     /// Report a fatal error with the current span.
     fn fatal(&self, &str) -> FatalError;
     /// Report a non-fatal error with the current span.
     fn err(&self, &str);
+    fn emit_fatal_errors(&mut self);
     fn peek(&self) -> TokenAndSpan;
     /// Get a token the parser cares about.
-    fn real_token(&mut self) -> TokenAndSpan {
-        let mut t = self.next_token();
+    fn try_real_token(&mut self) -> TokenAndSpanResult {
+        let mut t = self.try_next_token()?;
         loop {
             match t.tok {
                 token::Whitespace | token::Comment | token::Shebang(_) => {
-                    t = self.next_token();
+                    t = self.try_next_token()?;
                 }
                 _ => break,
             }
         }
-        t
+        Ok(t)
+    }
+    fn real_token(&mut self) -> TokenAndSpan {
+        match self.try_real_token() {
```

Comment:
could factor out the logic here and in next_token into its own method taking `self.try_real_token()` as a closure.